Bootstrap Email Editor [MAILPOET-5365]

costasovo opened the pull request:
## Description

This PR enables the creation of Standard newsletters using the Gutenberg email editor. The editor is built on top of the WP post editor, and email content is saved as a custom post type `mailpoet_email`. MailPoet-specific data are saved to the MailPoet newsletters table, but for now, it is only the association with the WP post.

Part of the PR is also integration into the plugin sending processes so the emails can be sent and previewed using a link from the Emails listing page. This should help us easily test sending and rendering. The email rendering is, for now, just a dummy class and will be addressed in the follow-up tickets.

For now, the editor allows just a few code blocks (paragraph, heading, columns) and the publish button was replaced by the "Next" button that takes a user to the current send page.

![Screenshot 2023-08-18 at 10 43 23](https://github.com/mailpoet/mailpoet/assets/1082140/d55255b2-e973-459c-b45d-804b52d6a1f9)


## Code review notes

Following the Automations example, both PHP and JS code are divided into `core` and `integration` subdirectories. Anything MailPoet specific is in the integration folder. For the core stuff, I avoid using other MailPoet-specific services and modules. The code in the core folder should work only with WP code or other stuff from the core.

Please review commit by commit, and feel free to ping me any time. I tried to add explanations to commit messages or code comments.

## QA notes

There is not much to test for now, as it is WIP behind a feature flag. So it is more about making yourself familiar with what is going on. There are a few changes in the current email rendering and sending code, so I want to ask you to check email previews and sending with the current editor. Acceptance tests are passing, so I assume some visual checks of sent emails should be enough.


To enable the editor, you need to go to /wp-admin/admin.php?page=mailpoet-experimental

The editor can be used only with standard newsletters. When you attempt to create a standard newsletter, you will be asked which editor you want to use.


## Linked PRs

_N/A_

## Linked tickets

[MAILPOET-5365]

## After-merge notes

_N/A_

## Tasks

- [x] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [x] I added sufficient test coverage
- [x] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes


[MAILPOET-5365]: https://mailpoet.atlassian.net/browse/MAILPOET-5365?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

costasovo committed:
Add Gutenberg email editor feature flag

[MAILPOET-5365]

costasovo committed:
Add EmailEditor class with basic registration of CPT for emails

The idea is that there is a core class
Mailpoet\MailPoet\EmailEditor\Core\EmailEditor which will contain common email editor related stuff
that can be used and shared between projects.
Mailpoet\MailPoet\EmailEditor\Integration\EmailEditor which contains MailPoet specific settings and and extra modifications
specific for MailPoet.
[MAILPOET-5365]

costasovo committed:
Restrict allowed blocks to paragraph, heading and columns

[MAILPOET-5365]

costasovo committed:
Add DB migration and wpPostId to NewsletterEntity

The email body will be stored as a wp_post, but for mailpoet specific attributes
we will use the current DB structure so that we can easy integrate emails edited in the new editor.
[MAILPOET-5365]

costasovo committed:
Save association to WP post with newsletter

This commit adds a simple implementation of ensuring the mailpoet_email post type
has associated NewsletterEntity.
In the future we will need to add support also for saving additional data for
newsletters (subject, sending settings).
[MAILPOET-5365]

costasovo committed:
Delete WP Posts when deleting Newsletter Entities

[MAILPOET-5365]

costasovo committed:
Duplicate associated post when duplicating newsletter

[MAILPOET-5365]

costasovo committed:
Setup JS assets entry point and build for the email editor

[MAILPOET-5365]

costasovo committed:
Add custom send button next to publish button

[MAILPOET-5365]

costasovo committed:
Add EmailApiController that adds MailPoet specific data to post api

When loading an email post in the editor we need to attach MailPoet
specific data like associated newsletter entity id, email type etc.

This is done using register_rest_field function.
So far we only send the associated entity ID but there is also a method `saveEmailData` that will
be used to save MailPoet specific email data.

[MAILPOET-5365]

costasovo committed:
Fetch MailPoet email data in editor and use id for redirect to send page

[MAILPOET-5365]

costasovo committed:
Update WebPack config for email editor to be based on @wordpress/scripts config

This brings us closer to how Gutenberg packages are build
and we can start writing CSS the same way as they are used in Gutenberg.
[MAILPOET-5365]

costasovo committed:
Add CSS for hiding publish button and enqueue CSS file for loading

[MAILPOET-5365]

costasovo committed:
Open standard emails created in the new editor in the new editor

[MAILPOET-5365]

costasovo committed:
Disable email content validation on send page for new editor

[MAILPOET-5365]

costasovo committed:
Update go back to design page link on send page for new editor

[MAILPOET-5365]

costasovo committed:
Refactor email body rendering into a separate class

[MAILPOET-5365]

costasovo committed:
Add dummy email body renderer for new editor

[MAILPOET-5365]

costasovo committed:
Rename send button to next button

[MAILPOET-5365]

costasovo committed:
Add modal for selecting email editor for standard newsletter

[MAILPOET-5365]

costasovo committed:
Add cleanup for registered block editor assets

[MAILPOET-5365]

costasovo committed:
Temporarily skip validation of email created via Gutenberg

[MAILPOET-5365]

costasovo committed:
Add hook for modifying editor settings

We will need to disable/change various editor settings.
This will be done in this method or via filter in the integration code.
[MAILPOET-5365]

costasovo committed:
Temporarily disable patterns

We will reintroduce block patterns post MVP.
We may want to filter or add patterns.
[MAILPOET-5365]

costasovo committed:
Refactor assets cleanup to an extra class

[MAILPOET-5365]

costasovo committed:
Save email draft on click to next

[MAILPOET-5365]

costasovo committed:
Introduce core directory in the email editor assets

The directory will be used for components and other JS assets that are meant for
the code that is intended to be part of possible future email editor package
that could be used in different projects.
[MAILPOET-5365]

costasovo committed:
Refactor Next button in email editor into a component

[MAILPOET-5365]

costasovo committed:
Add loading animation to the next button

[MAILPOET-5365]

costasovo committed:
Hardcode email width in the editor

This commit sets editor are width to 660px (based on the old editor).
[MAILPOET-5365]

costasovo committed:
Highlight correct submenu item when editing email

[MAILPOET-5365]

costasovo committed:
Add acceptance test for sending newsletter created in Gutenberg editor

[MAILPOET-5365]

costasovo committed:
Prevent displaying post editor welcome guide

This is rather an edge case when users open the email editor
before they have opened the post editor.
It also happens in acceptance tests.
[MAILPOET-5365]

costasovo committed:
Fix issue with saving session in acceptance tests for newer chrome versions

Using /dev/shm:/dev/shm as volume was not working when using a newer chrome
version on MAC.

[MAILPOET-5365]

costasovo committed:
Do not save newsletter templates when for Gutenberg emails

This will be disabled to prevent saving empty templates.
We will decide how/if we support templates in the future.
[MAILPOET-5365]

costasovo committed:
Adjust Gutenberg email editor test for iframed editor

[MAILPOET-5365]

lysyjan commented on mailpoet/lib/EmailEditor/Integration/EmailEditor.php:
The commit message confused me a little because it contains the class from the Core namespace twice. I understood because we discussed this code personally.

lysyjan commented on mailpoet/lib/EmailEditor/Core/EmailEditor.php:
```diff
@@ -0,0 +1,43 @@
+    $this->registerEmailPostTypes();
+  }
+
+  private function registerEmailPostTypes() {
```
We could add here a method description containing why this method registers our custom post type. Also, I noticed a missing return type, probably `void`.

lysyjan reviewed (dismissed):
Great job! You did a huge work here.
I'm looking forward to the next steps.

I think my two small remarks can be skipped because it's a first PR from many.
The only thing that I wanted to ask is if we should follow more the Automation directory pattern. It could make easier orientation for all developers across those two projects.

costasovo committed:
Refactor email editor PHP and JS code to follow automations namespacing

[MAILPOET-5365]

costasovo committed:
Add description to EmailEditor::registerEmailPostTypes

[MAILPOET-5365]

costasovo committed:
Add README.md for email editor

[MAILPOET-5365]

costasovo commented:
@lysyjan Thank you for the review and suggestions! I added three new commits and updated the confusing commit message.

Aschepikov commented:
Hi @costasovo  I have a fatal error when I am trying to open the Gutenberg editor, can you check it pls. If you need I can give a creds for my site
![image](https://github.com/mailpoet/mailpoet/assets/113185437/a81af869-e07c-4796-ae00-71688578c772)

![image](https://github.com/mailpoet/mailpoet/assets/113185437/dfa0b7e5-09ae-4c75-bf89-9b6e015a9bda)

costasovo committed:
Fix checking hooks registered using a class name in assets cleaner

The code was not expecting that the filter callback could be defined with a class name as a string.
[MAILPOET-5365]

costasovo commented:
@lysyjan I added one commit addressing the issue reported by Alexey.

Aschepikov reviewed (approved):
GJ!
